Improve internationalization of document types, payment and dispatch methods

niklasbuechner opened the pull request:
### 1. Why is this change necessary?
1. Because payments and their translations are different for English backend and frontend users.
![screen shot 2017-10-20 at 18 31 10](https://user-images.githubusercontent.com/31041526/31831900-58d848a4-b5c5-11e7-8e3a-6c818eff0043.png)
![screen shot 2017-10-20 at 18 31 18](https://user-images.githubusercontent.com/31041526/31831901-58f1dd0a-b5c5-11e7-881c-b01a4ef46118.png)
![screen shot 2017-10-20 at 18 31 37](https://user-images.githubusercontent.com/31041526/31831902-5909d7c0-b5c5-11e7-8151-a105f41078cf.png)


2. The configuration window for payment methods does not show the translated names.
3. There is a mix of German and English for payment methods
![screen shot 2017-10-20 at 18 37 09](https://user-images.githubusercontent.com/31041526/31832055-d7460384-b5c5-11e7-9a4c-45f7b5da48d2.png)


4. Shipping methods are not translated in the backend
5. Translations of document types are not shown
6. Payments and shipping methods are translated using snippets in the backend and using an `s_core_translations` entry in the frontend.

### 2. What does this change do, exactly?
The change extends the translation component of Shopware to provide a generic translation method for associative arrays. The translation methods can process any associative array if its translations are saved in `s_core_translations`. Furthermore, the service now provides small helper methods to translate payment means, document types and shipping methods.

Additionally, the corresponding backend components have been adjusted to use the new translation methods. This leads to payments, shipping and document information being translated throughout the whole backend, leading to a better internationalisation of the Shopware backend.


### 3. Describe each step to reproduce the issue or behaviour.
#### Backend:
1. Login with an English backend user
2. Open one of the following windows:
- Order listing (and its filters)
- Order detail view
- Order document view
- Customer data view
- Customer order list
- Payment configuration dialog
- Shipping methods config view
- PDF document creation list in basic settings
- Risk management

#### Frontend:
- Open the order list of a customer
- Create a new order and change the payment options

### 4. Please link to the relevant issues (if any).
None

### 5. Which documentation changes (if any) need to be made because of this PR?
None

### 6. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

xrogers committed:
SW-20476 - Correctly create QueryBuilder with $indexBy parameter

Fixes #1247

niklasbuechner committed:
Improve internationalization

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -31,6 +31,11 @@
+     */
+    const CONTEXT_BACKEND = 0;
+    const CONTEXT_FRONTEND = 1;
+    /**
```
Messing newline.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+     * @param $orders
+     * @param $context
+     * @param $language integer Language id
+     * @param $fallback integer Id of fallback language
```
The PHPdoc `@param`s should start with their typehint (e.g. `int`), followed by the parameter name and an optional description

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+        $dispatchTypes = [];
+
+        // Extract documents, payment and dispatch methods
+        for ($i = 0; $i < count($orders); $i += 1) {
```
Please use a `foreach` loop.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+            $orderDocuments = $orders[$i]['documents'];
+            $dispatchTypes[$orders[$i]['dispatch']['id']] = $orders[$i]['dispatch'];
+            $paymentTypes[$orders[$i]['payment']['id']] = $orders[$i]['payment'];
+            for ($orderCounter = 0; $orderCounter < count($orderDocuments); $orderCounter += 1) {
```
Please use a `foreach` loop.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+            $paymentTypes[$orders[$i]['payment']['id']] = $orders[$i]['payment'];
+            for ($orderCounter = 0; $orderCounter < count($orderDocuments); $orderCounter += 1) {
+                $type = $orderDocuments[$orderCounter]['type'];
+                $documentTypes[$type['id']] = $type;
```
This can be a single line.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+        $translationMapping = [['source' => 'dispatch_name', 'destination' => 'name']];
+
+        return $this->translateObjects(
+            $dispatches, 'config_dispatch', $context, $translationMapping, $language, $fallback
```
Please either move all parameters to the first line or separate them to one parameter per line.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+     * @param $language integer Language id
+     * @param $fallback integer Id of fallback language
+     *
+     * @return mixed Translated dispatches
```
Is the result really `mixed`?

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+    {
+        $translationMapping = [
+            'description',
+            ['source' => 'additionalDescription', 'destination' => 'additionaldescription'],
```
Please move the array elements to separate lines.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+
+        // Create correct mapping array from parameter
+        $translationMapping = [];
+        for ($i = 0; $i < count($translationMappingArg); $i += 1) {
```
Please use `array_map` here.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+        return $objects;
+    }
+
+    public function getLanguageIdsByContext($context)
```
Missing documentation comment.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+        return $languages;
+    }
+
+    public function getFallbackLocaleId($currentLocaleId)
```
Missing documentation comment.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+                $languages['fallback'] = $this->getFallbackLocaleId($languages['main']) ?: $languages['fallback'];
+                break;
+            case self::CONTEXT_FRONTEND:
+                $shop = Shopware()->Container()->get('shopware_storefront.context_service')->getShopContext()->getShop();
```
The services used in this components should be injected in the constructor.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+
+        switch ($context) {
+            case self::CONTEXT_BACKEND:
+                $identity = Shopware()->Container()->get('Auth')->getIdentity();
```
The services used in this components should be injected in the constructor.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +396,295 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+            return 1;
+        }
+
+        $fallback = (int) $this->connection->fetchColumn(
```
Please use `intval()` when returning the value instead of casting it.

svenmuennich commented on engine/Shopware/Controllers/Backend/Base.php:
```diff
@@ -146,6 +146,11 @@ public function getPaymentsAction()
 
+        // translate payments
+        /** @var $translationComponent Shopware_Components_Translation */
+        $translationComponent = Shopware()->Container()->get('translation');
```
There's no need to access the DI container (`Shopware()->Container()`) directly, since any controller has a `get()` method for loading services.

svenmuennich reviewed (changes requested):
I didn't duplicate all my feedback to all lines it applies to. So please keep that in mind when changing the code to check all modified files for similar changes 😉

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +397,327 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+     * and translates them.
+     *
+     * @param array $orders
+     * @param int   $context
```
Please don't indent the parameter names, because keeping the indentation up to date in future changes might clutter the diff (since all rows must be updated).

niklasbuechner replied:
As discussed personally, the indentation was created by shopware's git hook.

niklasbuechner replied:
As discussed personally, the indentation was created by shopware's git hook.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +397,327 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+            return [
+                    'source' => $value,
+                    'destination' => $value,
+                ];
```
Indentation...

niklasbuechner replied:
As discussed personally, the indentation was created by shopware's git hook.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +36,48 @@ class Shopware_Components_Translation
     /**
+     * @var int
+     */
+    private $language;
```
I think this property should be called `localeId`.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +36,48 @@ class Shopware_Components_Translation
+    /**
+     * @var int
+     */
+    private $fallback;
```
And this one `fallbackLocaleId`.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +36,48 @@ class Shopware_Components_Translation
+     * @param Container|null  $container
      */
-    public function __construct(Connection $connection = null)
+    public function __construct(Connection $connection = null, $container = null)
```
Missing typehint for `$container`.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +36,48 @@ class Shopware_Components_Translation
+
+        // Determine how to query the current language
+        if ($container->has('Auth')) {
+            $locale = $identity = Shopware()->Container()->get('Auth')->getIdentity()->locale;
```
Please use `$container` instead of `Shopware()->Container()`.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +36,48 @@ class Shopware_Components_Translation
+            }
+        } elseif ($container->has('Shop')) {
+            $localeName = $container->get('Shop')->getLocale()->getLocale();
+            $locale = Shopware()->Models()->getRepository('Shopware\Models\Shop\Locale')->findOneBy([
```
Please use `$container` instead of `Shopware()->Models()`.

svenmuennich replied:
Also, why do we query the locale here again. We already have it as an object in the shop, don't we?

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +36,48 @@ class Shopware_Components_Translation
+        $this->connection = $connection ?: $container->get('dbal_connection');
+
+        // Determine how to query the current language
+        if ($container->has('Auth')) {
```
Please use the exact name of the services as defined in `services.xml` etc., because the DI container sometimes gets confused when getting the same service using a different name (capitalization) and creates multiple instances of the same service.

niklasbuechner replied:
That is the exact name. However, it does not refer to a service (defined in a service.xml) but rather to a Shopware default plugin.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +36,48 @@ class Shopware_Components_Translation
+            }
+        }
+
+        if (!$this->language) {
```
We need a comment here explaining why we fall back to locale ID `1`.

svenmuennich commented on engine/Shopware/Components/Translation.php:
```diff
@@ -391,6 +428,288 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+        }
+
+        $fallback = $this->connection->fetchColumn(
+            "SELECT id FROM s_core_locales WHERE locale = 'en_GB'"
```
What happens, if the `en_GB` locale does not exist?

niklasbuechner replied:
Added a fallback for the fallback to German 😉

DirkPersky committed:
SW-20419 - Added og:description & twitter:description to custom and form
pages
fixes 1367

wolv-dev committed:
SW-20499 - Fix not changing tab on mobile devices when clicking on rating link

Fixes: https://github.com/shopware/shopware/pull/1370

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +37,46 @@ class Shopware_Components_Translation
+        }
+
+        if (!$this->localeId) {
+            // Fallback to german language in case no local exists
```
s/german language/default locale/
s/local/locale/

fixpunkt replied:
Still not fixed.

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -36,11 +37,46 @@ class Shopware_Components_Translation
+
+        // Determine how to query the current language
+        if ($container->has('Auth')) {
+            $locale = $identity = $container->get('Auth')->getIdentity()->locale;
```
Uh, I think you can remove the `$identity = ` part?

fixpunkt replied:
This is still not fixed.

Commit:
SW-20486 - [BUGFIX] Prevent Fatal error when links or download is null
fixes 1374

fixpunkt reviewed (changes requested):
Partial review only, sry :/

wolv-dev committed:
SW-20500 - [TASK] Changed translation of newsletter in footer and added shopName config

Fixes #1375

kekster committed:
SW-20545 - fixed generating emotion landingpage seo urls
fixes 1385

Scribblerockerz committed:
SW-20572 - Add unique placeholders in condition parser trait

Fixes: https://github.com/shopware/shopware/pull/1394

Commit:
SW-20584 - Workaround for MariaDB >= 10.2.7 compatibility

Attribute default models are broken otherwise, see: https://github.com/doctrine/doctrine2/issues/6565
fixes 1396

resslinger committed:
SW-20600 - fix emotion banner component
fixes #1403

Commit:
SW-17153 - Added all attributes to product export

wolv-dev committed:
SW-20684 - [BUGFIX] Fixed ordering of blog entries in shopping world element
fixes #1417

Commit:
SW-20660 - use typecast instead of intval, strval, doubleval and floatval
fixes #1420

jkrzefski committed:
SW-20661 - Add filter event in sOrder::sSaveOrder for mail context

Fixes: https://github.com/shopware/shopware/pull/1421

uehler committed:
SW-20667 - fix product layout in manufacturer action
fixes #1423

soebbing committed:
SW-19592 - Add style fixes

soebbing committed:
SW-19592 - Add flag to prevent images in albums to be garbage collected

Commit:
SW-20701 - Remove unnecessary method getBlogCategoryListIds

Closes #1427

janbuecker committed:
Simplify docker setup

janbuecker committed:
Disable unstable mink test

Commit:
SW-20718 - fixed nonstatic call of static method
fixes #1433

Commit:
NTR - Nightwatch.js for E2E tests

JoshuaBehrens committed:
SW-20736 - Remove duplicate catch block

Fixes: https://github.com/shopware/shopware/pull/1434

OliverSkroblin committed:
SW-20545 - Replace double quotes with single quotes

janbuecker committed:
SW-20684 - Check existing of blog in fetched array and add tests

janbuecker committed:
SW-20476 - Add tests for ModelRepository

janbuecker committed:
Update composer dependencies

Commit:
NTR - Add documentation and add the ability to provide env arguments to the runner

janbuecker committed:
Change nightwatch artifacts path

janbuecker committed:
Add bamboo build scripts for nightwatch

janbuecker committed:
Elasticsearch integration

janbuecker committed:
Update elasticsearch/elasticsearch to support authentication

soebbing committed:
SW-20739 - Add general delay of 750 to remote email checks

soebbing committed:
SW-19780 - Don't show a canonical if the page has noindex

jenskueper committed:
SW-20499 - Optimize rating/note buttons for mobile

philipreinken committed:
SW-19219 - Add conditional statements to temporarily skip an ajax call on customer registration in certain cases.

Commit:
SW-20679 - Hide the listing button beneath emotions when category has no products

janbuecker committed:
Add node_modules to .gitignore

soebbing committed:
Fix translation in newsletter sender column title

taltholtmann committed:
SW-19013 - Add detail tabs to window management panel

soebbing committed:
SW-20924 - Fix listing widget pagination

philipreinken committed:
SW-19969 - Add deprecation note for the articleId column in the s_article_attributes table

Commit:
SW-20951 - Correct statistics plugin when nosniff header is enabled

philipreinken committed:
SW-19715 - Display an error message when a request during search indexing fails

soebbing committed:
Merge pull request #5892 in SW/shopware from sw-20584/5.3/auto-imported-from-github to 5.3

* commit '20131c375d7d923ec18b932ea7f5300d688ee497':
  SW-20584 - Workaround for MariaDB >= 10.2.7 compatibility

janbuecker committed:
Merge pull request #5961 in SW/shopware from ntr/5.3/simplify-docker-setup to 5.3

* commit 'bfea3b544b00602156738d043cecc9d5fcd60f4e':
  Add node_modules to .gitignore
  Update elasticsearch/elasticsearch to support authentication
  Elasticsearch integration
  Add bamboo build scripts for nightwatch
  Change nightwatch artifacts path
  NTR - Add documentation and add the ability to provide env arguments to the runner
  NTR - Nightwatch.js for E2E tests
  Update composer dependencies
  Simplify docker setup
  Disable unstable mink test

soebbing committed:
Merge pull request #6011 in SW/shopware from sw-19780/5.3/hide-canonical-on-noindex-pages to 5.3

* commit 'cbfb54c3d4163ebaef37653e82534c473420202b':
  SW-19780 - Don't show a canonical if the page has noindex

soebbing committed:
Merge pull request #6084 in SW/shopware from sw-19969/5.3/add-deprecation-note to 5.3

* commit '7e37f3099915ffaafd5027d387aa38a8748cd02e':
  SW-19969 - Add deprecation note for the articleId column in the s_article_attributes table

soebbing committed:
Merge pull request #6008 in SW/shopware from sw-20739/5.3/add-email-check-delay to 5.3

* commit '734acb18dfb179053a87bffc5cdf8d33769e9465':
  SW-20739 - Add general delay of 750 to remote email checks

soebbing committed:
Merge pull request #6031 in SW/shopware from sw-20679/5.3/hide-listing-button-empty-categories to 5.3

* commit '7cb68228d2574b1066d8657713e31a48ffc40d1a':
  SW-20679 - Hide the listing button beneath emotions when category has no products

soebbing committed:
Merge pull request #5954 in SW/shopware from sw-19592/5.3/media-garbage-collection to 5.3

* commit '1ff9147434123d1222b7a1c91e5aef82dc6a7da7':
  SW-19592 - Add flag to prevent images in albums to be garbage collected
  SW-19592 - Add style fixes

Commit:
SW-15287 - Fixed the _private folder not loading on checkout finish

janbuecker committed:
Merge pull request #6095 in SW/shopware from sw-15287/5.3/fixed-smarty-modifier-on-finish to 5.3

* commit '056488f8a75790c6470e679ffd461740ae2ef309':
  SW-15287 - Fixed the _private folder not loading on checkout finish

janbuecker committed:
Merge pull request #6096 in SW/shopware from github-christmas-run to 5.3

* commit 'cd2e02d44586e1c4ae7c105d61e5e4f5af07d8ff':
  Fix translation in newsletter sender column title
  SW-20499 - Optimize rating/note buttons for mobile
  SW-20572 - Add unique placeholders in condition parser trait
  SW-20667 - fix product layout in manufacturer action fixes #1423
  SW-20476 - Add tests for ModelRepository
  SW-20476 - Correctly create QueryBuilder with $indexBy parameter
  SW-20500 - [TASK] Changed translation of newsletter in footer and added shopName config
  SW-20661 - Add filter event in sOrder::sSaveOrder for mail context
  SW-20419 - Added og:description & twitter:description to custom and form pages fixes 1367
  SW-20684 - Check existing of blog in fetched array and add tests
  SW-20499 - Fix not changing tab on mobile devices when clicking on rating link
  SW-20736 - Remove duplicate catch block
  SW-20684 - [BUGFIX] Fixed ordering of blog entries in shopping world element fixes #1417
  SW-20545 - Replace double quotes with single quotes
  SW-20701 - Remove unnecessary method getBlogCategoryListIds
  SW-20718 - fixed nonstatic call of static method fixes #1433
  SW-20660 - use typecast instead of intval, strval, doubleval and floatval fixes #1420
  SW-20600 - fix emotion banner component fixes #1403
  SW-20545 - fixed generating emotion landingpage seo urls fixes 1385
  SW-20486 - [BUGFIX] Prevent Fatal error when links or download is null fixes 1374

fixpunkt commented on _sql/demo/latest.sql:
```diff
@@ -6129,7 +6129,8 @@ INSERT IGNORE INTO `s_core_translations` (`id`, `objecttype`, `objectdata`, `obj
 (263, 'article', 'a:2:{s:10:"txtArtikel";s:24:"Shipping costs by weight";s:19:"txtlangbeschreibung";s:226:"<p>This article weighs 8 KG. These requires a calculation by weight. If you add it to the cart, it will be charged with 19,90 Euro. To do so, use the follwing access data: max.mustermann@mail.com &amp; Password: shopware .</p>";}', 248, '2'),
 (270, 'article', 'a:1:{s:10:"txtArtikel";s:19:"Beach Towel "Ibiza"";}', 178, '2'),
-(272, 'config_payment', 'a:5:{i:4;a:2:{s:11:"description";s:7:"Invoice";s:21:"additionalDescription";s:141:"Payment by invoice. Shopware provides automatic invoicing for all customers on orders after the first, in order to avoid defaults on payment.";}i:2;a:2:{s:11:"description";s:5:"Debit";s:21:"additionalDescription";s:15:"Additional text";}i:3;a:2:{s:11:"description";s:16:"Cash on delivery";s: ...
+(272, 'config_payment', 'a:5:{i:4;a:2:{s:11:"description";s:7:"Invoice";s:21:"additionalDescription";s:141:"Payment by invoice. Shopware provides automatic invoicing for all customers on orders after the first, in order to avoid defaults on payment.";}i:2;a:2:{s:11:"description";s:5:"Debit";s:21:"additionalDescription";s:15:"Additional text";}i:3;a:2:{s:11:"description";s:16:"Cash on delivery";s: ...
```
Based on the review comments from the last review, we probably shouldn't modify any of the `latest.sql` files.

niklasbuechner replied:
This is only an issue for `install/latest.sql`, so `demo/latest.sql` should be fine.

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +426,289 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
     }
 
+    /**
+     * Helper method to translate orders.
```
Redundant

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +426,289 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
 
+    /**
+     * Helper method to translate orders.
+     * Method extracts all documents, payment methods and dispatch methods of the order
```
"Method" is unnecessary here.

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +426,289 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+     * @param int   $language
+     * @param int   $fallback
+     *
+     * @return mixed
```
Could be more precise

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +426,289 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+    public function translateOrders($orders, $language = null, $fallback = null)
+    {
+        $documentTypes = [];
+        $paymentTypes = [];
```
`paymentMethods`

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +426,289 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+    {
+        $documentTypes = [];
+        $paymentTypes = [];
+        $dispatchTypes = [];
```
`dispatchMethods`

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +426,289 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+    }
+
+    /**
+     * Helper method to translate dispatches.
```
"dispatch methods".

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +426,289 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+    }
+
+    /**
+     * Configurable function to translate associative arrays and map the translations to
```
Holy shit, this is wordy.

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +426,289 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+     *
+     * @return array Translated objects
+     */
+    public function translateObjects($objects, $type, $translationMappingArg, $language = null, $fallback = null)
```
Wow, this method is quite complicated. Let's try to make it work kind of like this (from the client perspective):

```php
$translator = $this->getObjectTranslator('config_payment', $language, $fallback);

        $translatedPayments = [];
        foreach ($payments as $payment) {
            $translatedPayments[] = $translator->translateObjectProperty($payment, 'additionaldescription', 'additionalDescription', $fallback);
        }

        $translationMapping = [
            'description',
            [
                'source' => 'additionalDescription',
                'destination' => 'additionaldescription',
            ],
        ];
```

soebbing committed:
Merge pull request #6083 in SW/shopware from sw-20924/5.3/fix-listing-widget-pagination to 5.3

* commit '9d046e9c091459e5ca372b334bdf58a0f1c3de2e':
  SW-20924 - Fix listing widget pagination

soebbing committed:
Merge pull request #6019 in SW/shopware from sw-19219/5.3/confirmation-behaviour to 5.3

* commit '09205444916b6004db5ebd45b4b61a0544aff700':
  SW-19219 - Add conditional statements to temporarily skip an ajax call on customer registration in certain cases.

fixpunkt commented on tests/Functional/Controllers/Backend/OrderTest.php:
```diff
@@ -118,6 +118,164 @@ public function testBatchProcessOrderDocument()
+        $response = $this->dispatch('backend/Order/getList?' . $getString);
+
+        $str = $response->getBody();
+        $responseStr = preg_replace('/(new Date\([-0-9]+\))/', '"$1"', $response->getBody());
```
This line needs a comment ;)

fixpunkt reviewed (changes requested):
Great job!

I think we could improve the title a bit. "Correct" seems to be the wrong verb here, since that's mostly used as in "To correct someone". Also, "Backend translations" is not specific enough - it could refer to any i18n happening in the backend. I'd propose something like "Translate configured entities in the backend consistently", but maybe you have another good idea.

We also need to try and simplify the PR a bit, as we've just discussed face-to-face.

soebbing committed:
Merge pull request #6072 in SW/shopware from sw-19013/5.3/Add-customer-detail-tab-to-window-management-panel to 5.3

* commit '37f97cf04625664ee3ac49918efbd4744abcb72d':
  SW-19013 - Add detail tabs to window management panel

ndzoesch committed:
SW-21005 - Add CLA.md
fixes #1480

soebbing committed:
Merge pull request #6098 in SW/shopware from sw-21005/5.3/auto-imported-from-github to 5.3

* commit '4049db978b1170d07f4bcf14e3c8f263a3913e35':
  SW-21005 - Add CLA.md fixes #1480

soebbing committed:
SW-20745 - Make email in order clickable

Phil23 committed:
Merge pull request #6101 in SW/shopware from sw-20745/5.3/make-email-clickable-again to 5.3

* commit 'e38860c259d27cb30879bbd2580f3a6a80a8a785':
  SW-20745 - Make email in order clickable

soebbing committed:
SW-20739 - Fix lagging email input

Phil23 committed:
Merge pull request #6102 in SW/shopware from sw-20739/5.3/fix-lagging-email-input to 5.3

* commit '7ec3d4610d77a4aef0e09ee11110565fe31a399f':
  SW-20739 - Fix lagging email input

soebbing committed:
Merge pull request #6086 in SW/shopware from sw-20951/5.3/nosniff-header-on-statistics-plugin to 5.3

* commit 'f1d6c5d4a96b092db5172918807ed578d0d16506':
  SW-20951 - Correct statistics plugin when nosniff header is enabled

soebbing committed:
Merge pull request #5926 in SW/shopware from sw-17153/5.3/add-all-attributes-in-product-export to 5.3

* commit '1f2aca988934797ae0a0a1fab618e0204e3ca2e7':
  SW-17153 - Added all attributes to product export

soebbing committed:
Merge pull request #6116 in SW/shopware from sw-19715/5.3/search-index-error-message to 5.3

* commit '1585cf4d6d4f371ff34144b2e3044f6f5eda049f':
  SW-19715 - Display an error message when a request during search indexing fails

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -35,12 +37,47 @@ class Shopware_Components_Translation
+    public function __construct(Connection $connection = null, Container $container = null)
     {
-        $this->connection = $connection ?: Shopware()->Container()->get('dbal_connection');
+        $container = ($container) ?: Shopware()->Container();
```
lose the parens.

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +427,174 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
     }
 
+    /**
+     * Helper to translate orders.
```
This sentence is redundant. Also, please don't use "Method" or "Class" as a subject when documenting classes or methods, since that's just irrelevant. Correct phpdoc for this method would be:

> Translates a list of orders by translating each order's documents, payment methods and dispatch methods.
>
> @param [...]

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +427,174 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+
+        // Translate the objects
+        $translatedDocumentTypes = $this->translateDocuments($documentTypes, $language, $fallback);
+        $translatedDispatchTypes = $this->translateDispatchMethods($dispatchMethods, $language, $fallback);
```
*Methods

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +427,174 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+        // Translate the objects
+        $translatedDocumentTypes = $this->translateDocuments($documentTypes, $language, $fallback);
+        $translatedDispatchTypes = $this->translateDispatchMethods($dispatchMethods, $language, $fallback);
+        $translatedPaymentTypes = $this->translatePaymentMethods($paymentMethods, $language, $fallback);
```
*Methods

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +427,174 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+    }
+
+    /**
+     * Helper to translate dispatch methods.
```
lose the "Helper"

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +427,174 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+    }
+
+    /**
+     * Factory method to create a properly configured object translator
```
"Creates an object translator for a specific type of translatable object."

fixpunkt commented on engine/Shopware/Components/Translation.php:
```diff
@@ -390,6 +427,174 @@ public function write($language, $type, $key = 1, $data = null, $merge = false)
+    }
+
+    /**
+     * Loads the id of the fallback language
```
.

fixpunkt commented on engine/Shopware/Components/Translation/ObjectTranslator.php:
```diff
@@ -0,0 +1,97 @@
+
+namespace Shopware\Components\Translation;
+
+class ObjectTranslator
```
This should have a short class description explaining what this object's purpose in life is.

fixpunkt commented on engine/Shopware/Components/Translation/ObjectTranslator.php:
```diff
@@ -0,0 +1,97 @@
+
+class ObjectTranslator
+{
+    private $fallback;
```
Please add phpdoc for all of these properties. The phpdoc should at the very least specify the types of the properties.

fixpunkt commented on engine/Shopware/Controllers/Backend/Base.php:
```diff
@@ -257,6 +262,10 @@ public function getDispatchesAction()
         //select all shop as array
         $data = $query->getArrayResult();
 
+        // translate dispatches
```
*methods

fixpunkt commented on engine/Shopware/Controllers/Backend/Config.php:
```diff
@@ -287,6 +287,18 @@ public function getListAction()
         }
 
+        if ($name === 'document') {
+            // Translate documents
```
Doesn't actually translate documents, maybe "Translates document type names" is more precise?

fixpunkt commented on engine/Shopware/Controllers/Backend/Shipping.php:
```diff
@@ -78,6 +78,17 @@ public function getShippingCostsAction()
         $shippingCosts = $paginator->getIterator()->getArrayCopy();
         $shippingCosts = $this->convertShippingCostsDates($shippingCosts);
 
+        // Translate shipping cost
```
Probably needs an update

niklasbuechner committed:
Adjust tests, migration and UPGRADE.md

fixpunkt commented:
@niklasbuechner can you link the PR against https://github.com/shopware/shopware here and close this PR please?
